### PR TITLE
Add out wrappers to some decompositions

### PR DESCRIPTION
**Pull request by isuruf**

### CODE REVIEW

**Review by lezcano — COMMENTED** [review]
out_wrapper should be after `pw_cast`. Otherwise this looks good, the errors may be coming from incorrect samples or some xfails that may need removing?

File: torch/_decomp/decompositions.py [hunk 1]
```diff
@@ -2528,8 +2532,9 @@ def _upsample_nearest2d_common(input, h_indices, w_indices):
     return result
 
 
-@register_decomposition(aten.upsample_nearest2d.default)
+@register_decomposition([aten.upsample_nearest2d.default, aten.upsample_nearest2d.out])
 @aten.upsample_nearest2d.default.py_impl(DispatchKey.Autograd)
+@out_wrapper()
```
Comment (peterbell10) [inline torch/_decomp/decompositions.py:2537]: Test failures are real, it's expecting channels last input to give a channels last output when passing in a zero-size out tensor.
https://github.com/pytorch/pytorch/blob/310f6ab11a4b36fddab0bb309566d52834bff0f5/test/test_torch.py#L8048

But in `out_wrapper` it just resizes without using any specific memory format, 
https://github.com/pytorch/pytorch/blob/310f6ab11a4b36fddab0bb309566d52834bff0f5/torch/_prims_common/wrappers.py#L175

I'm guessing this is a problem for other decompositions as well, but this is the only function that tests for it.

Comment (peterbell10) [reply torch/_decomp/decompositions.py:2537]: Looks like for `TensorIterator` the behavior is to give the resized tensor the same strides as the functional output, but I expect this isn't true for all operators and some likely just call `resize_`.

Comment (lezcano) [reply torch/_decomp/decompositions.py:2537]: We should then add a optional parameter to `out_wrapper` that implements this logic

**Comment by lezcano** [comment]
This PR is quite reasonable and shouldn't be difficult to finish. Would be good to get it through the finish line at some point.

File: torch/_prims_common/wrappers.py [hunk 2]
```diff
@@ -170,9 +170,9 @@ def _resize_output_check(out: TensorLikeType, shape: ShapeType):
 
 
 # TODO: handle tuples of tensors
-def _maybe_resize_out(out: TensorLikeType, shape: ShapeType):
+def _maybe_resize_out(out: TensorLikeType, shape: ShapeType, memory_format: Optional[torch.memory_format]):
```
Comment (peterbell10) [inline torch/_prims_common/wrappers.py:173]:
```suggestion
def _maybe_resize_out(out: TensorLikeType, shape: ShapeType, memory_format: Optional[torch.memory_format] = None):
```

**Review by peterbell10 — COMMENTED** [review]
Note this will need a rebase after #122783 is merged

File: torch/_prims_common/wrappers.py [hunk 3]
```diff
@@ -218,6 +227,15 @@ def out_wrapper(*out_names: str, exact_dtype: bool = False, pass_is_out: bool =
...
     is_tensor = len(out_names) == 1
 
+    def compute_memory_format(t):
+        if (
+            preserve_memory_format
+            and utils.suggest_memory_format(t) == torch.channels_last
+        ):
+            return torch.channels_last
+        else:
+            return None
```
Comment (lezcano) [inline torch/_prims_common/wrappers.py:237]: Shouldn't this be
```suggestion
    def maybe_compute_memory_format(t):
        return utils.suggest_memory_format(t) if preserve_memory_format else None
```

**Review by lezcano — APPROVED** [review]
Great! Thank you!

**Pull request closed without merging**